adam/learner/affordances.py

```diff
@@ -233,3 +248,94 @@ def _fallback_templates(
                         if isinstance(node, ObjectSemanticNodePerceptionPredicate)
                     }
                 ), 1.0
+
+
+@attrs(slots=True)
+class MappingAffordanceLearner(AbstractAffordanceTemplateLearner):
+
+    concept_to_affordances: MutableMapping[Concept, MutableSet[Concept]] = attrib(
+        validator=deep_mapping(instance_of(Concept), deep_iterable(instance_of(Concept))),
+        factory=dict,
+    )
+    debug_to_concept: MutableMapping[str, Concept] = attrib(
+        validator=deep_mapping(instance_of(str), instance_of(Concept)), factory=dict
+    )
+
+    def _match_template(
+        self,
+        *,
+        concept: Concept,
+        pattern: PerceptionGraphTemplate,
+        perception_graph: PerceptionGraph,
+        confidence: float,
+    ) -> Iterable[Tuple[PerceptionGraphPatternMatch, SemanticNode]]:
+        return immutableset()
+
+    def _new_concept(self, debug_string: str) -> Concept:
+        return AffordanceConcept(debug_string)
+
+    def _learning_step(
+        self,
+        language_perception_semantic_alignment: LanguagePerceptionSemanticAlignment,
+        bound_surface_template: SurfaceTemplateBoundToSemanticNodes,
+    ) -> None:
+        for _, object_semantic in bound_surface_template.slot_to_semantic_node.items():
+            debug_name = bound_surface_template.surface_template.to_short_string()
+            if debug_name in self.debug_to_concept:
+                concept = self.debug_to_concept[debug_name]
+            else:
+                concept = self._new_concept(debug_name)
+                self.debug_to_concept[debug_name] = concept
+
+            if object_semantic.concept not in self.concept_to_affordances:
+                self.concept_to_affordances[object_semantic.concept] = set()
+            if concept not in self.concept_to_affordances[object_semantic.concept]:
+                self.concept_to_affordances[object_semantic.concept].add(concept)
```

Comment:
I think you can write this slightly more concisely as:

```suggestion
            self.concept_to_affordances.setdefault(
                object_semantic.concept, set()
            ).add(concept)
```

Reply:
TIL: Dictionaries have a `setdefault` method.